src/v/cloud_topics/level_zero/stm/ctp_stm_state.cc

```diff
@@ -12,105 +12,57 @@
+    _projected_epoch = std::max(
+      epoch, _projected_epoch.value_or(cluster_epoch{}));
 }
 
-bool ctp_stm_state::snapshot_exists(dl_snapshot_id id) const noexcept {
-    return std::binary_search(
-      _snapshots.begin(),
-      _snapshots.end(),
-      id,
-      [](const dl_snapshot_id& a, const dl_snapshot_id& b) {
-          return a.version < b.version;
-      });
+std::optional<kafka::offset>
+ctp_stm_state::get_last_reconciled_offset() const noexcept {
+    return _last_reconciled_offset;
 }
 
-std::optional<dl_snapshot_payload>
-ctp_stm_state::read_snapshot(dl_snapshot_id id) const {
-    auto it = std::find_if(
-      _snapshots.begin(), _snapshots.end(), [&id](const dl_snapshot_id& s) {
-          return s.version == id.version;
-      });
-
-    // Snapshot not found.
-    if (it == _snapshots.end()) {
-        return std::nullopt;
-    }
-
-    return dl_snapshot_payload{
-      .id = *it,
-    };
+std::optional<model::offset>
+ctp_stm_state::get_last_reconciled_log_offset() const noexcept {
+    return _last_reconciled_log_offset;
 }
 
-void ctp_stm_state::remove_snapshots_before(dl_version last_version_to_keep) {
-    if (_snapshots.empty()) {
-        throw std::runtime_error(fmt::format(
-          "Attempt to remove snapshots before version {} but no snapshots "
-          "exist",
-          last_version_to_keep));
-    }
-
-    // Find the first snapshot to keep. It is the first snapshot with a version
-    // equal or greater than the version to keep.
-    auto it = std::lower_bound(
-      _snapshots.begin(),
-      _snapshots.end(),
-      last_version_to_keep,
-      [](const dl_snapshot_id& a, dl_version b) { return a.version < b; });
-
-    if (it == _snapshots.begin()) {
-        // Short circuit if there are no snapshots to remove
-        return;
-    } else if (it == _snapshots.end()) {
-        throw std::runtime_error(fmt::format(
-          "Trying to remove snapshots before an non-existent snapshot",
-          last_version_to_keep));
-    } else {
-        _snapshots.erase(_snapshots.begin(), it);
-    }
+void ctp_stm_state::advance_epoch(cluster_epoch epoch) {
+    // Propagate projected epoch on a follower
+    _projected_epoch = std::max(
+      _projected_epoch.value_or(cluster_epoch{}), epoch);
+    // Register new epoch
+    _max_epoch = std::max(epoch, _max_epoch.value_or(cluster_epoch{}));
 }
 
-ctp_stm_offsets& ctp_stm_state::get_offsets() noexcept { return _offsets; }
-
-const ctp_stm_offsets& ctp_stm_state::get_offsets() const noexcept {
-    return _offsets;
+void ctp_stm_state::advance_last_reconciled_offset(
+  kafka::offset new_last_reconciled_offset,
+  model::offset new_last_reconciled_log_offset) noexcept {
+    _last_reconciled_offset = std::max(
+      _last_reconciled_offset.value_or(kafka::offset{}),
+      new_last_reconciled_offset);
+    _last_reconciled_log_offset = std::max(
+      _last_reconciled_log_offset.value_or(model::offset{}),
+      new_last_reconciled_log_offset);
```

Comment:
The epochs and the _log_ variant of LRO seem to be internally calculated, so instead of advancing here with `max` i'd expect verification that they are non-decreasing. the kafka LRO comes from external to the system, so races means that we might have out-of-order values and max makes more sense?